### PR TITLE
[SYNCOPE 660] Asynchronous jobs delegated to Quartz, like Task and Report, now can be checked and interrupted during execution

### PR DESCRIPTION
**Pull request by giacomolm**

### CODE REVIEW

File: common/src/main/java/org/apache/syncope/common/services/ReportService.java [hunk 1]
```diff
@@ -192,4 +195,25 @@
     @Consumes({ MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON })
     Response exportExecutionResult(@NotNull @PathParam("executionId") Long executionId,
             @QueryParam("format") ReportExecExportFormat fmt);
+    
+    
+    /**
+     * List all the reports of a particular type
```
Comment (ilgrosso) [inline common/src/main/java/org/apache/syncope/common/services/ReportService.java]: Please change this description: it should be something like as "List report jobs of the given type"

File: common/src/main/java/org/apache/syncope/common/services/TaskService.java [hunk 2]
```diff
@@ -242,4 +244,24 @@ TaskExecTO execute(@NotNull @PathParam("taskId") Long taskId,
     @Produces({ MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON })
     @Consumes({ MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON })
     BulkActionResult bulk(@NotNull BulkAction bulkAction);
+
+    /**
+     * List all the tasks of a particular type
```
Comment (ilgrosso) [inline common/src/main/java/org/apache/syncope/common/services/TaskService.java]: Please change this description: it should be something like as "List task jobs of the given type"

File: common/src/main/java/org/apache/syncope/common/services/ReportService.java [hunk 3]
```diff
@@ -192,4 +195,25 @@
...
+     *
+     * @param type of report (like All, Running, Scheduled)
```
Comment (ilgrosso) [inline common/src/main/java/org/apache/syncope/common/services/ReportService.java]: Same here: "type of report job" (no need to give examples, you have a proper Enum)

File: common/src/main/java/org/apache/syncope/common/types/JobAction.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * under the License.
+ */
+package org.apache.syncope.common.types;
+
+import javax.xml.bind.annotation.XmlEnum;
+
+@XmlEnum
+public enum JobAction {
+
+    START("start"),
```
Comment (ilgrosso) [inline common/src/main/java/org/apache/syncope/common/types/JobAction.java]: No need for providing the `fromString` or explicit labels here (same for all enums)

File: core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java [hunk 5]
```diff
@@ -99,6 +100,11 @@
     protected Task task;
 
     /**
+     * The current running thread containing the task to be executed.
+     */
+    AtomicReference<Thread> runningThread = new AtomicReference<Thread>();
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java]: Please explicitly set the scope here (hint: `protected`)

File: core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java [hunk 6]
```diff
@@ -176,4 +183,15 @@ public void execute(final JobExecutionContext context) throws JobExecutionExcept
...
     }
+
+    @Override
+    public void interrupt() throws UnableToInterruptJobException {        
+        Thread thread = this.runningThread.getAndSet(null);
+        if (thread != null) {
+            LOG.info("Interrupting job time {} ", new Date().toString());
+            thread.interrupt();
+        }
+        else LOG.debug("Unable to retrieve the right thread related to the current job execution");
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java]: Please add brackets.

File: core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java [hunk 7]
```diff
@@ -176,4 +183,15 @@ public void execute(final JobExecutionContext context) throws JobExecutionExcept
     protected boolean hasToBeRegistered(final TaskExec execution) {
         return false;
     }
+
+    @Override
+    public void interrupt() throws UnableToInterruptJobException {        
+        Thread thread = this.runningThread.getAndSet(null);
+        if (thread != null) {
+            LOG.info("Interrupting job time {} ", new Date().toString());
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/quartz/AbstractTaskJob.java]: Please consistently use `LOG.debug` and format the date according to `SyncopeConstants` format.

File: core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                }
+                break;
+            case RUNNING:
+                try {
+                    for (JobExecutionContext jec : scheduler.getScheduler().getCurrentlyExecutingJobs()) {
+                        Long jobId = getIdFromJobName(jec.getJobDetail().getKey());
+                        if (jobId != null) {
+                            E jobExecTO = reference.newInstance();
+                            jobExecTO.setId(jobId);
+                            jobExecTO.setStatus("Running");
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java]: Why not using `scheduler.getScheduler().getTriggerState(t.getKey()).name()`?

File: core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                    for (String groupName : scheduler.getScheduler().getJobGroupNames()) {
+                        for (JobKey jobKey : scheduler.getScheduler().getJobKeys(GroupMatcher.jobGroupEquals(groupName))) {
+                            Long jobId = getIdFromJobName(jobKey);
+                            if (jobId != null) {
+                                List<? extends Trigger> jobTriggers = scheduler.getScheduler().getTriggersOfJob(jobKey);
+                                for (Trigger t : jobTriggers) {
+                                    E jobExecTO = reference.newInstance();
+                                    jobExecTO.setId(jobId);
+                                    jobExecTO.setStatus(scheduler.getScheduler().getTriggerState(t.getKey()).name());
+                                    jobExecTO.setStartDate(t.getStartTime());
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java]: Why not setting status here?

File: core/src/main/java/org/apache/syncope/core/rest/controller/ReportController.java [hunk 10]
```diff
@@ -341,4 +340,24 @@ protected ReportTO resolveReference(final Method method, final Object... args)
 
         throw new UnresolvedReferenceException();
     }
+
+    @Override
+    @PreAuthorize("hasRole('REPORT_READ')")
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/rest/controller/ReportController.java]: Shouldn't this be `REPORT_LIST`?

File: core/src/main/java/org/apache/syncope/core/rest/controller/TaskController.java [hunk 11]
```diff
@@ -401,4 +404,26 @@ protected AbstractTaskTO resolveReference(final Method method, final Object... a
 
         throw new UnresolvedReferenceException();
     }
+
+    @Override
+    @PreAuthorize("hasRole('TASK_READ')")
```
Comment (ilgrosso) [inline core/src/main/java/org/apache/syncope/core/rest/controller/TaskController.java]: Same as above: `TASK_LIST`

**Comment by ilgrosso** [comment]
Some minor comments in-line, nice work @giacomolm LGTM!

File: core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java (hunk 8 above)
Comment (giacomolm) [inline core/src/main/java/org/apache/syncope/core/rest/controller/AbstractJobController.java]: This should works ```scheduler.getScheduler().getTriggerState(jec.getTrigger().getKey()).name()```

**Comment by giacomolm** [comment]
thanks @ilgrosso !! Please check this new revision c1faf13

**Comment by ilgrosso** [comment]
All is fine @giacomolm, please go ahead and merge

**Pull request merged**